<a href="https://colab.research.google.com/github/AmnaNoor123/urdu-ocr-codesaviours-si26-amna/blob/main/SI26_Week2_amna.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**This notebook covers Week 2:**

Image Preprocessing for the Urdu OCR project.

The images we collected vary in size, color, and pixel quality. Before training any model, we need to standardize them — this is called preprocessing. It ensures the model can learn smoothly from consistent, clean input rather than noisy, inconsistent raw data.

In this notebook, we will:

Convert images to grayscale
Resize all images to a standard dimension
Remove noise for cleaner text
Binarize the images (pure black/white pixels)
Test Tesseract OCR on the processed Urdu images to see where it fails

In [13]:
from google.colab import drive
drive.mount('/content/drive')

import csv, os

synthetic_labels = [
    {'image': 'data/raw/Synthetic raw images/urdu_1.png', 'text': 'پاکستان زندہ باد'},
    {'image': 'data/raw/Synthetic raw images/urdu_2.png', 'text': 'آج کا موسم خوشگوار ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_3.png', 'text': 'تعلیم ہر انسان کا حق ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_4.png', 'text': 'کراچی پاکستان کا سب سے بڑا شہر ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_5.png', 'text': 'محنت کامیابی کی کنجی ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_6.png', 'text': 'کتاب علم کا خزانہ ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_7.png', 'text': 'وقت کی قدر کرنی چاہیے'},
    {'image': 'data/raw/Synthetic raw images/urdu_8.png', 'text': 'صحت ہزار نعمت ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_9.png', 'text': 'پانی زندگی کے لیے ضروری ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_10.png', 'text': 'سچائی ہمیشہ کامیاب ہوتی ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_11.png', 'text': 'لاہور ایک تاریخی شہر ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_12.png', 'text': 'بچوں کو اچھی تربیت دینی چاہیے'},
    {'image': 'data/raw/Synthetic raw images/urdu_13.png', 'text': 'دوستی ایک قیمتی رشتہ ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_14.png', 'text': 'صبر کا پھل میٹھا ہوتا ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_15.png', 'text': 'اردو ایک خوبصورت زبان ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_16.png', 'text': 'درخت لگانا ضروری ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_17.png', 'text': 'والدین کی عزت کرنی چاہیے'},
    {'image': 'data/raw/Synthetic raw images/urdu_18.png', 'text': 'علم حاصل کرنے کی کوئی عمر نہیں ہوتی'},
    {'image': 'data/raw/Synthetic raw images/urdu_19.png', 'text': 'ایمانداری بہترین پالیسی ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_20.png', 'text': 'پاکستان کا موسم بہت متنوع ہے'},
]

synthetic_naskh_labels = [
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_1.png', 'text': 'دریا کنارے خوبصورت منظر تھا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_2.png', 'text': 'بچے پارک میں کھیل رہے ہیں'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_3.png', 'text': 'استاد نے سبق اچھی طرح سمجھایا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_4.png', 'text': 'بازار میں بہت رش تھا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_5.png', 'text': 'میں نے کتاب پڑھ کر رکھ دی'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_6.png', 'text': 'چاند رات کو بہت روشن تھا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_7.png', 'text': 'کسان کھیت میں کام کر رہا تھا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_8.png', 'text': 'دوست نے مجھے تحفہ دیا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_9.png', 'text': 'بارش کے بعد ہوا خوشگوار ہو گئی'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_10.png', 'text': 'ماں نے کھانا بنایا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_11.png', 'text': 'طالب علم امتحان کی تیاری کر رہا ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_12.png', 'text': 'پرندے صبح سویرے چہچہاتے ہیں'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_13.png', 'text': 'شہر کی سڑکیں کشادہ ہیں'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_14.png', 'text': 'پہاڑوں پر برف پڑی ہوئی تھی'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_15.png', 'text': 'دکاندار نے سامان تول کر دیا'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_16.png', 'text': 'علم انسان کی سب سے بڑی طاقت ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_17.png', 'text': 'محنت کبھی رائیگاں نہیں جاتی'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_18.png', 'text': 'صحت مند جسم میں صحت مند دماغ ہوتا ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_19.png', 'text': 'پاکستان کے شمالی علاقے بہت خوبصورت ہیں'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_20.png', 'text': 'کتاب انسان کی بہترین دوست ہے'},
    {'image': 'data/raw/Synthetic raw images/urdu_naskh_21.png', 'text': 'زندگی میں صبر بہت ضروری ہے'},
]

handwritten_labels = [
    {'image': 'data/raw/Handwritten/handwritten_1.jpeg', 'text': 'میرے کپڑے استری کردو'},
    {'image': 'data/raw/Handwritten/handwritten_2.jpeg', 'text': 'مجھے کھیلنا پسند ہے'},
    {'image': 'data/raw/Handwritten/handwritten_3.jpeg', 'text': 'ٹماٹر، پیاز'},
    {'image': 'data/raw/Handwritten/handwritten_4.jpeg', 'text': 'میں کھانا گرم کرکے آتی ہوں'},
    {'image': 'data/raw/Handwritten/handwritten_5.jpeg', 'text': 'میں ایک مجبور لڑکی ہوں'},
    {'image': 'data/raw/Handwritten/handwritten_6.jpeg', 'text': 'میں کام کرنا جانتی ہوں'},
    {'image': 'data/raw/Handwritten/handwritten_7.jpeg', 'text': 'میں ایک سندر لڑکی ہوں'},
    {'image': 'data/raw/Handwritten/handwritten_8.jpeg', 'text': 'سیب کا ڈالیہ دے دو'},
    {'image': 'data/raw/Handwritten/handwritten_9.jpeg', 'text': 'مجھے کھانا پکانے کا شوق ہے'},
    {'image': 'data/raw/Handwritten/handwritten_10.jpeg', 'text': 'میں بہت سمجھدار لڑکی ہوں'},
    {'image': 'data/raw/Handwritten/handwritten_11.jpeg', 'text': 'کھانا'},
    {'image': 'data/raw/Handwritten/handwritten_12.jpeg', 'text': 'آلو'},
    {'image': 'data/raw/Handwritten/handwritten_13.jpeg', 'text': 'میں ایک خوبصورت لڑکی ہوں'},
    {'image': 'data/raw/Handwritten/handwritten_14.jpeg', 'text': 'میرا نام آمنہ ہے'},
    {'image': 'data/raw/Handwritten/handwritten_15.jpeg', 'text': 'میں گانا گانا جانتی ہوں'},
]

newspaper_labels = [
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_183748.png', 'text': 'خواہش!'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184136.png', 'text': 'تازہ ترین خبروں، تجزیے اور تبصرے کے لیے یہاں کلک کیجیے'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184154.png', 'text': 'ٹی ایل پی سربراہ سعد رضوی اور ان کے بھائی کی گمشدگی کا معمہ برقرار'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184212.png', 'text': 'ایران نے مذاکرات کے لیے رابطہ کیا ہے مگر اس سے پہلے کارروائی بھی ہوسکتی ہے، ٹرمپ'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184222.png', 'text': 'ملک میں شدید سردی کی افواہوں کی تردید'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184703.png', 'text': 'ہر دفعہ پتھر کو لوگ دل سمجھ لیتے ہیں، عمریں بیت جاتی ہیں دل کو دل بنانے میں'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184759.png', 'text': 'کچھ تعلقات ختم کرنے ہی کو پڑتے ہیں'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184814.png', 'text': 'اس وقت پاکستان بہت سے ملکی اور بین الاقوامی مسائل میں گھرا ہوا ہے'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184838.png', 'text': 'علم بڑی دولت ہے، ہم اسے حاصل کرنے کے لیے صرف محنت کرتے ہیں'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184851.png', 'text': 'پاکستان بنیادی طور پر ایک زرعی ملک ہے'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_184947.png', 'text': 'وقت کی پابندی'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185017.png', 'text': 'ہر لذت کی انتہا بیزاری ہے'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185028.png', 'text': 'اختلاف کے باوجود احترام سے پیش آنا کمزوری نہیں'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185046.png', 'text': 'راز، تم مرد ہو کر عورت کو بدنام کرتے ہو'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185057.png', 'text': 'انسان مکان بدلتا ہے رشتے بدلتا ہے'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185144.png', 'text': 'غلطیاں بھی ہوں گی، غلط بھی سمجھا جانے گا یہ زندگی ہے'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185221.png', 'text': 'درخت، مشکل، یونٹ، جب تک، معمول'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185235.png', 'text': 'ہجر'},
    {'image': 'data/raw/Newspaper/Screenshot_2026-07-21_185309.png', 'text': 'چاہت'},
    {'image': 'data/raw/Newspaper/WhatsApp_Image_2026-07-21_at_18_34_40.jpeg', 'text': 'بارش کی تباہ کاریاں، 16 افراد پانی میں بہہ گئے'},
    {'image': 'data/raw/Newspaper/0.png', 'text': 'پشاور،بنوں (نمائندہ جنگ،اے ایف پی) بنوں میں اقوام میراخیل اور'},
    {'image': 'data/raw/Newspaper/1.png', 'text': 'اسکے ساتھ ملحقہ علاقے کے عوام نے بجلی و گیس'},
    {'image': 'data/raw/Newspaper/9.png', 'text': 'میں اقوام میراخیل اور اسکے ساتھ ملحقہ علاقے کے عوام'},
    {'image': 'data/raw/Newspaper/10.png', 'text': 'نے بجلی وگیس کی لوڈشیڈنگ کیخلاف مظاہرہ کیا۔ مظاہرین نے'},
    {'image': 'data/raw/Newspaper/17.png', 'text': 'ملانے والی مرکزی شاہراہ پر کرفیو نافذ کردیا گیا ہے۔بنوں'},
    {'image': 'data/raw/Newspaper/100.png', 'text': 'نہیں آ یا۔ آخریہ کونسا مذہب ہے؟کیا اسلام کے نام'},
    {'image': 'data/raw/Newspaper/1000.png', 'text': 'ضرور ہے۔ انہوں نے کہا کہ میرا احمدیوں سے کوئی'},
    {'image': 'data/raw/Newspaper/1001.png', 'text': 'تعلق نہیں، انہوں نے کہا کہ پولیس کو اپنی کارکردگی'},
    {'image': 'data/raw/Newspaper/6651.png', 'text': 'میں کُل جماعتی کشمیر رابطہ کمیٹی کے زیر اہتمام ہونے'},
    {'image': 'data/raw/Newspaper/6652.png', 'text': 'والے کشمیر سیمینار کے ساتھ بھر پو تعاون کا فیصلہ'},
    {'image': 'data/raw/Newspaper/6653.png', 'text': 'کیا گیا۔ چوہدری ظفر اقبال نے کہا لبریشن لیگ کے'},
    {'image': 'data/raw/Newspaper/6654.png', 'text': 'سربراہ کے ایچ خورشید مرحوم نے آج سے نصف صدی'},
    {'image': 'data/raw/Newspaper/6655.png', 'text': 'پہلے کشمیریوں اور پاکستانیوں کو جو تجاویز دیں وہ اگر'},
    {'image': 'data/raw/Newspaper/6656.png', 'text': 'اپنائی جاتیں تو یہ مسئلہ کب کا حل ہوجاتا۔ محمد'},
    {'image': 'data/raw/Newspaper/6657.png', 'text': 'ایاز نے خطاب کرتے ہوئے کہا کہ آزاد کشمیر میں'},
    {'image': 'data/raw/Newspaper/6658.png', 'text': 'لبریشن لیگ واحد جماعت ہے جس نے اصول کی سیاست'},
    {'image': 'data/raw/Newspaper/6659.png', 'text': 'کی ہے، مقبوضہ کشمیر میں ہونے والے مظالم کو دنیا'},
    {'image': 'data/raw/Newspaper/8816.png', 'text': 'حد سے بڑھیں تو پاکستان میں پہلی مرتبہ 1953ء میں'},
    {'image': 'data/raw/Newspaper/8817.png', 'text': 'قادیانیت کے خلاف تحریک چلی، علماء اہلسنّت نے مسئلہ ختم'},
    {'image': 'data/raw/Newspaper/8818.png', 'text': 'نبوت اور فتنہ قادیانیت کو طشت ازبام کیا جبکہ دوسری'},
    {'image': 'data/raw/Newspaper/8819.png', 'text': 'دفعہ 1974ء میں قادیانیت کے خلاف تحریک چلی اور کامیابی'},
    {'image': 'data/raw/Newspaper/8820.png', 'text': 'سے ہمکنار ہوئی، انہوں نے کہا کہ غیر مسلم اقلیت'},
    {'image': 'data/raw/Newspaper/10000.png', 'text': 'ڈالر کی قدر میں اضافہ ہوگیا تاہم یورپین ممالک کی'},
    {'image': 'data/raw/Newspaper/10001.png', 'text': 'اہم کرنسیوں بشمول یورو اور برطانوی پاؤنڈ کی قدر میں'},
    {'image': 'data/raw/Newspaper/10002.png', 'text': 'کمی ریکارڈ کی گئی۔ منگل کو امریکی ڈالر کی قیمت'},
    {'image': 'data/raw/Newspaper/10003.png', 'text': 'فروخت اور قیمت خرید میں بالترتیب 21 پیسے اور 20'},
    {'image': 'data/raw/Newspaper/10004.png', 'text': 'پیسے کی تیزی رہی، تاہم یورو کی قیمت فروخت میں'},
    {'image': 'data/raw/Newspaper/10005.png', 'text': '50 پیسے اور قیمت خرید میں 49 پیسے کی کمی'},
    {'image': 'data/raw/Newspaper/10006.png', 'text': 'ریکارڈ کی گئی۔ اسی طرح برطانوی پاؤنڈ کی قدر میں'},
    {'image': 'data/raw/Newspaper/10007.png', 'text': 'بھی 45 پیسے کی مندی رہی۔ نیشنل بینک آف پاکستان'},
    {'image': 'data/raw/Newspaper/10008.png', 'text': 'سے جاری کردہ اعداد و شمار کے مطابق امریکی ڈالر'},
    {'image': 'data/raw/Newspaper/10009.png', 'text': 'اپنی گزشتہ کاروباری روز کی قیمت فروخت 88.17 روپے سے'},
]

signboards_labels = [
    {'image': 'data/raw/Sign boards/Screenshot_2026-07-21_185327.png', 'text': 'ارمان'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_56_04.jpeg', 'text': 'سرٹیفکیٹ کے حصول کے لیے درخواست'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_56_14.jpeg', 'text': 'وہاں ریاض مسلسل سے کام چلتا ہے، یہاں گلے کے سہارے کلام چلتا ہے'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_56_40.jpeg', 'text': 'کرکٹ کے کھیل کا موازنہ مشاعرے کے ساتھ کیا گیا ہے'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_57_07.jpeg', 'text': 'کرکٹ اور مشاعرہ (نظم) از دلاور فگار'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_58_13.jpeg', 'text': 'شیطان آپ کو ہر وقت غمگین کیوں کرتا ہے'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_58_25.jpeg', 'text': 'تیسرے حرف کے مطابق: اگر تیسرے حرف پر فتحہ یا کسرہ ہو تو ہمزہ وصلی کو کسرہ دیں گے'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_58_37.jpeg', 'text': 'تیسرے حرف کے مطابق: اگر تیسرے حرف پر ضمہ ہو تو ہمزہ وصلی کو ضمہ دیں گے'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_19_00_40.jpeg', 'text': 'میں سالار سکندر سے بات کرنا چاہتی ہوں، آپ امامہ ہاشم ہیں؟'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_19_01_02.jpeg', 'text': 'جو مری ہے وہ میری بیوی تھی اور مجھے بہت پیاری تھی'},
    {'image': 'data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_19_01_14.jpeg', 'text': 'اعمال کے نتائج ہوتے ہیں اور وہ بھگتنے بھی پڑتے ہیں'},
    {'image': 'data/raw/Sign boards/ء_998.png', 'text': 'ء'},
    {'image': 'data/raw/Sign boards/ء_999.png', 'text': 'ء'},
    {'image': 'data/raw/Sign boards/ا_973.png', 'text': 'ا'},
    {'image': 'data/raw/Sign boards/ا_974.png', 'text': 'ا'},
    {'image': 'data/raw/Sign boards/آ_993.png', 'text': 'آ'},
    {'image': 'data/raw/Sign boards/آ_994.png', 'text': 'آ'},
    {'image': 'data/raw/Sign boards/ب_996.png', 'text': 'ب'},
    {'image': 'data/raw/Sign boards/ب_997.png', 'text': 'ب'},
    {'image': 'data/raw/Sign boards/پ_94.png', 'text': 'پ'},
]

books_labels = [
    {'image': 'data/raw/Books/ۃ_96.png', 'text': 'ۃ'},
    {'image': 'data/raw/Books/ۃ_97.png', 'text': 'ۃ'},
    {'image': 'data/raw/Books/ت_95.png', 'text': 'ت'},
    {'image': 'data/raw/Books/ت_96.png', 'text': 'ت'},
    {'image': 'data/raw/Books/ٹ_98.png', 'text': 'ٹ'},
    {'image': 'data/raw/Books/ٹ_99.png', 'text': 'ٹ'},
    {'image': 'data/raw/Books/ث_983.png', 'text': 'ث'},
    {'image': 'data/raw/Books/ث_984.png', 'text': 'ث'},
    {'image': 'data/raw/Books/ج_97.png', 'text': 'ج'},
    {'image': 'data/raw/Books/ج_98.png', 'text': 'ج'},
    {'image': 'data/raw/Books/چ_95.png', 'text': 'چ'},
    {'image': 'data/raw/Books/چ_96.png', 'text': 'چ'},
    {'image': 'data/raw/Books/ح_970.png', 'text': 'ح'},
    {'image': 'data/raw/Books/ح_971.png', 'text': 'ح'},
    {'image': 'data/raw/Books/خ_949.png', 'text': 'خ'},
    {'image': 'data/raw/Books/خ_950.png', 'text': 'خ'},
    {'image': 'data/raw/Books/د_95.png', 'text': 'د'},
    {'image': 'data/raw/Books/د_96.png', 'text': 'د'},
    {'image': 'data/raw/Books/ڈ_95.png', 'text': 'ڈ'},
    {'image': 'data/raw/Books/ڈ_96.png', 'text': 'ڈ'},
    {'image': 'data/raw/Books/ذ_98.png', 'text': 'ذ'},
    {'image': 'data/raw/Books/ذ_99.png', 'text': 'ذ'},
    {'image': 'data/raw/Books/ر_980.png', 'text': 'ر'},
    {'image': 'data/raw/Books/ر_981.png', 'text': 'ر'},
    {'image': 'data/raw/Books/ڑ_93.png', 'text': 'ڑ'},
    {'image': 'data/raw/Books/ڑ_94.png', 'text': 'ڑ'},
    {'image': 'data/raw/Books/ز_99.png', 'text': 'ز'},
    {'image': 'data/raw/Books/ز_970.png', 'text': 'ز'},
    {'image': 'data/raw/Books/ژ_98.png', 'text': 'ژ'},
    {'image': 'data/raw/Books/ژ_99.png', 'text': 'ژ'},
    {'image': 'data/raw/Books/س_97.png', 'text': 'س'},
    {'image': 'data/raw/Books/س_98.png', 'text': 'س'},
    {'image': 'data/raw/Books/ش_983.png', 'text': 'ش'},
    {'image': 'data/raw/Books/ش_984.png', 'text': 'ش'},
    {'image': 'data/raw/Books/ص_98.png', 'text': 'ص'},
    {'image': 'data/raw/Books/ص_99.png', 'text': 'ص'},
    {'image': 'data/raw/Books/ض_979.png', 'text': 'ض'},
    {'image': 'data/raw/Books/ض_980.png', 'text': 'ض'},
    {'image': 'data/raw/Books/ط_973.png', 'text': 'ط'},
    {'image': 'data/raw/Books/ط_974.png', 'text': 'ط'},
]

other_labels = [
    {'image': 'data/raw/Other/ظ_99.png', 'text': 'ظ'},
    {'image': 'data/raw/Other/ظ_987.png', 'text': 'ظ'},
    {'image': 'data/raw/Other/ع_99.png', 'text': 'ع'},
    {'image': 'data/raw/Other/ع_990.png', 'text': 'ع'},
    {'image': 'data/raw/Other/غ_99.png', 'text': 'غ'},
    {'image': 'data/raw/Other/غ_984.png', 'text': 'غ'},
    {'image': 'data/raw/Other/ف_993.png', 'text': 'ف'},
    {'image': 'data/raw/Other/ف_994.png', 'text': 'ف'},
    {'image': 'data/raw/Other/ق_992.png', 'text': 'ق'},
    {'image': 'data/raw/Other/ق_993.png', 'text': 'ق'},
    {'image': 'data/raw/Other/ک_0.png', 'text': 'ک'},
    {'image': 'data/raw/Other/ک_1.png', 'text': 'ک'},
    {'image': 'data/raw/Other/گ_0.png', 'text': 'گ'},
    {'image': 'data/raw/Other/گ_1.png', 'text': 'گ'},
    {'image': 'data/raw/Other/ل_983.png', 'text': 'ل'},
    {'image': 'data/raw/Other/ل_984.png', 'text': 'ل'},
    {'image': 'data/raw/Other/م_95.png', 'text': 'م'},
    {'image': 'data/raw/Other/م_96.png', 'text': 'م'},
    {'image': 'data/raw/Other/م_97.png', 'text': 'م'},
    {'image': 'data/raw/Other/م_98.png', 'text': 'م'},
    {'image': 'data/raw/Other/ن_96.png', 'text': 'ن'},
    {'image': 'data/raw/Other/ن_97.png', 'text': 'ن'},
    {'image': 'data/raw/Other/ھ_0.png', 'text': 'ھ'},
    {'image': 'data/raw/Other/ھ_1.png', 'text': 'ھ'},
    {'image': 'data/raw/Other/ہ_96.png', 'text': 'ہ'},
    {'image': 'data/raw/Other/ہ_97.png', 'text': 'ہ'},
    {'image': 'data/raw/Other/و_0.png', 'text': 'و'},
    {'image': 'data/raw/Other/و_1.png', 'text': 'و'},
    {'image': 'data/raw/Other/ے_51.png', 'text': 'ے'},
    {'image': 'data/raw/Other/ے_52.png', 'text': 'ے'},
    {'image': 'data/raw/Other/ی_99.png', 'text': 'ی'},
    {'image': 'data/raw/Other/ی_983.png', 'text': 'ی'},
]

labels = (synthetic_labels + synthetic_naskh_labels + handwritten_labels +
          newspaper_labels + signboards_labels + books_labels + other_labels)

os.makedirs('/content/drive/MyDrive/UrduOCR', exist_ok=True)
with open('/content/drive/MyDrive/UrduOCR/labels.csv', 'w', newline='', encoding='utf-8') as f:
    writer = csv.DictWriter(f, fieldnames=['image', 'text'])
    writer.writeheader()
    writer.writerows(labels)

print(f'Done! labels.csv Drive pe seedha ban gayi with {len(labels)} entries')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Done! labels.csv Drive pe seedha ban gayi with 200 entries


In [14]:
import shutil, os
DRIVE_ROOT = '/content/drive/MyDrive/UrduOCR'
os.makedirs('data', exist_ok=True)

if os.path.exists(f'{DRIVE_ROOT}/raw'):
    shutil.copytree(f'{DRIVE_ROOT}/raw', 'data/raw', dirs_exist_ok=True)

shutil.copy(f'{DRIVE_ROOT}/labels.csv', 'data/labels.csv')

import pandas as pd
print('Labels restored:', len(pd.read_csv('data/labels.csv')))

Labels restored: 200


In [15]:
import zipfile
zip_path = '/content/drive/MyDrive/Books.zip'

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall('data/raw')

print('Unzipped folders:', os.listdir('data/raw'))

Unzipped folders: ['Newspaper', 'Books', 'Handwritten', 'Other', 'Sign boards', 'Synthetic raw images']


In [16]:
import re

def clean_filename(name):
    base, ext = os.path.splitext(name)
    def decode_match(m):
        return chr(int(m.group(1), 16))
    base = re.sub(r'#U([0-9a-fA-F]{4,6})', decode_match, base)
    ext = re.sub(r'#U([0-9a-fA-F]{4,6})', decode_match, ext)
    base = re.sub(r'\s*\(\d+\)$', '', base)
    full = base + ext
    full = re.sub(r'(\.\w+)\1$', r'\1', full)
    return full

renamed = 0
for root, dirs, files in os.walk('data/raw'):
    for fname in files:
        new_name = clean_filename(fname)
        if new_name != fname:
            os.rename(os.path.join(root, fname), os.path.join(root, new_name))
            renamed += 1

print(f'Renamed {renamed} files')

Renamed 20 files


Preprocessing

In [17]:
!pip install opencv-python-headless pillow
import cv2, glob

def preprocess_image(image_path, save_path):
    img = cv2.imread(image_path)
    if img is None:
        print(f'Could not load: {image_path}')
        return
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    resized = cv2.resize(gray, (512, 128))
    denoised = cv2.fastNlMeansDenoising(resized, h=10)
    _, binary = cv2.threshold(denoised, 127, 255, cv2.THRESH_BINARY)
    cv2.imwrite(save_path, binary)
    return binary

os.makedirs('data/processed', exist_ok=True)

all_images = glob.glob('data/raw/**/*.jpg', recursive=True)
all_images += glob.glob('data/raw/**/*.png', recursive=True)
all_images += glob.glob('data/raw/**/*.jpeg', recursive=True)

print(f'Found {len(all_images)} images to process')
processed_count = 0
for img_path in all_images:
    save_path = f'data/processed/{os.path.basename(img_path)}'
    if preprocess_image(img_path, save_path) is not None:
        processed_count += 1

print(f'Done! Processed {processed_count} images')

Found 200 images to process
Done! Processed 200 images


In [18]:
import re

def normalize(name):
    return re.sub(r'[\s_]+', '', name).lower()

actual_files = [f for f in glob.glob('data/raw/**/*', recursive=True) if os.path.isfile(f)]
lookup = {normalize(os.path.basename(f)): f for f in actual_files}

df = pd.read_csv('data/labels.csv')
fixed_rows, auto_fixed, still_missing = [], 0, []
for _, row in df.iterrows():
    path = row['image']
    if os.path.exists(path):
        fixed_rows.append({'image': path, 'text': row['text']})
        continue
    key = normalize(os.path.basename(path))
    if key in lookup:
        fixed_rows.append({'image': lookup[key], 'text': row['text']})
        auto_fixed += 1
    else:
        still_missing.append(path)

print(f'Total labels: {len(df)}')
print(f'Already correct: {len(fixed_rows) - auto_fixed}')
print(f'Auto-fixed: {auto_fixed}')
print(f'Still missing: {len(still_missing)}')
for m in still_missing:
    print('  -', m)

corrected_df = pd.DataFrame(fixed_rows)
corrected_df.to_csv('data/labels.csv', index=False, encoding='utf-8')
corrected_df.to_csv(f'{DRIVE_ROOT}/labels.csv', index=False, encoding='utf-8')
print(f'\nSaved: {len(corrected_df)} entries (local + Drive)')

Total labels: 200
Already correct: 169
Auto-fixed: 20
Still missing: 11
  - data/raw/Newspaper/WhatsApp_Image_2026-07-21_at_18_34_40.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_56_04.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_56_14.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_56_40.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_57_07.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_58_13.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_58_25.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_18_58_37.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_19_00_40.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_19_01_02.jpeg
  - data/raw/Sign boards/WhatsApp_Image_2026-07-21_at_19_01_14.jpeg

Saved: 189 entries (local + Drive)


In [19]:
import glob
print("Newspaper WhatsApp files:")
for f in glob.glob('data/raw/Newspaper/WhatsApp*'):
    print(' ', f)

print("\nSign boards WhatsApp files:")
for f in glob.glob('data/raw/Sign boards/WhatsApp*'):
    print(' ', f)

Newspaper WhatsApp files:
  data/raw/Newspaper/WhatsApp Image 2026-07-21 at 18.34.40.jpeg

Sign boards WhatsApp files:
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.58.25.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.58.13.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 19.01.02.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 19.01.14.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 19.00.40.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.56.04.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.57.07.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.56.40.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.58.37.jpeg
  data/raw/Sign boards/WhatsApp Image 2026-07-21 at 18.56.14.jpeg


In [20]:
def normalize(name):
    return re.sub(r'[^a-zA-Z0-9]+', '', name).lower()

actual_files = [f for f in glob.glob('data/raw/**/*', recursive=True) if os.path.isfile(f)]
lookup = {normalize(os.path.basename(f)): f for f in actual_files}

df = pd.read_csv('data/labels.csv')
fixed_rows, auto_fixed, still_missing = [], 0, []
for _, row in df.iterrows():
    path = row['image']
    if os.path.exists(path):
        fixed_rows.append({'image': path, 'text': row['text']})
        continue
    key = normalize(os.path.basename(path))
    if key in lookup:
        fixed_rows.append({'image': lookup[key], 'text': row['text']})
        auto_fixed += 1
    else:
        still_missing.append(path)

print(f'Total labels: {len(df)}')
print(f'Already correct: {len(fixed_rows) - auto_fixed}')
print(f'Auto-fixed: {auto_fixed}')
print(f'Still missing: {len(still_missing)}')
for m in still_missing:
    print('  -', m)

corrected_df = pd.DataFrame(fixed_rows)
corrected_df.to_csv('data/labels.csv', index=False, encoding='utf-8')
corrected_df.to_csv(f'{DRIVE_ROOT}/labels.csv', index=False, encoding='utf-8')
print(f'\nSaved: {len(corrected_df)} entries (local + Drive)')

Total labels: 189
Already correct: 189
Auto-fixed: 0
Still missing: 0

Saved: 189 entries (local + Drive)


In [21]:
from google.colab import files
uploaded = files.upload()

Saving generate_labels.py to generate_labels.py


In [22]:
%run /content/generate_labels.py

Done! labels.csv Drive pe seedha ban gayi with 200 entries


In [23]:
def normalize(name):
    return re.sub(r'[^a-zA-Z0-9]+', '', name).lower()

actual_files = [f for f in glob.glob('data/raw/**/*', recursive=True) if os.path.isfile(f)]
lookup = {normalize(os.path.basename(f)): f for f in actual_files}

df = pd.read_csv(f'{DRIVE_ROOT}/labels.csv')
fixed_rows, auto_fixed, still_missing = [], 0, []
for _, row in df.iterrows():
    path = row['image']
    if os.path.exists(path):
        fixed_rows.append({'image': path, 'text': row['text']})
        continue
    key = normalize(os.path.basename(path))
    if key in lookup:
        fixed_rows.append({'image': lookup[key], 'text': row['text']})
        auto_fixed += 1
    else:
        still_missing.append(path)

print(f'Total labels: {len(df)}')
print(f'Already correct: {len(fixed_rows) - auto_fixed}')
print(f'Auto-fixed: {auto_fixed}')
print(f'Still missing: {len(still_missing)}')
for m in still_missing:
    print('  -', m)

corrected_df = pd.DataFrame(fixed_rows)
corrected_df.to_csv('data/labels.csv', index=False, encoding='utf-8')
corrected_df.to_csv(f'{DRIVE_ROOT}/labels.csv', index=False, encoding='utf-8')
print(f'\nSaved: {len(corrected_df)} entries (local + Drive)')

Total labels: 200
Already correct: 169
Auto-fixed: 31
Still missing: 0

Saved: 200 entries (local + Drive)


In [24]:
!apt-get install -y tesseract-ocr tesseract-ocr-urd
!pip install pytesseract

import pytesseract
from PIL import Image

# 5 processed images pe test karo
test_images = list(glob.glob('data/processed/*.png'))[:5]

print('=== Tesseract Results on Urdu Images ===')
print()

for img_path in test_images:
    img = Image.open(img_path)
    result = pytesseract.image_to_string(img, lang='urd')
    print(f'Image: {img_path}')
    print(f'Tesseract output: {result}')
    print('---')

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
The following NEW packages will be installed:
  tesseract-ocr-urd
0 upgraded, 1 newly installed, 0 to remove and 3 not upgraded.
Need to get 1,000 kB of archives.
After this operation, 1,413 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tesseract-ocr-urd all 1:4.00~git30-7274cfa-1.1 [1,000 kB]
Fetched 1,000 kB in 0s (7,395 kB/s)
Selecting previously unselected package tesseract-ocr-urd.
(Reading database ... 118243 files and directories currently installed.)
Preparing to unpack .../tesseract-ocr-urd_1%3a4.00~git30-7274cfa-1.1_all.deb ...
Unpacking tesseract-ocr-urd (1:4.00~git30-7274cfa-1.1) ...
Setting up tesseract-ocr-urd (1:4.00~git30-7274cfa-1.1) ...
=== Tesseract Results on Urdu Images ===

Image: data/processed/م_96.png
Tesseract output: 
---
Image: data/processed/66

In [25]:
df = pd.read_csv('data/labels.csv')

comparison = []
for img_path in test_images:
    fname = os.path.basename(img_path)
    match = df[df['image'].str.contains(re.escape(fname), regex=True)]
    actual_text = match.iloc[0]['text'] if not match.empty else 'Not found in labels.csv'

    img = Image.open(img_path)
    result = pytesseract.image_to_string(img, lang='urd').strip()

    comparison.append((fname, actual_text, result))
    print(f'Image: {fname}')
    print(f'Actual text:      {actual_text}')
    print(f'Tesseract output: {result}')
    print('---')

Image: م_96.png
Actual text:      م
Tesseract output: 
---
Image: 6652.png
Actual text:      والے کشمیر سیمینار کے ساتھ بھر پو تعاون کا فیصلہ
Tesseract output: الات ناک مت گر اھ
---
Image: ذ_99.png
Actual text:      ذ
Tesseract output: الا
---
Image: Screenshot 2026-07-21 184154.png
Actual text:      ٹی ایل پی سربراہ سعد رضوی اور ان کے بھائی کی گمشدگی کا معمہ برقرار
Tesseract output: ٹی ئیل پی سریراہ سحد رضوی اور ات کے بھاکی کی
گمعدگی کا مسا یرظراں

یلت کا ہام

   

 

097ك:/
---
Image: 10003.png
Actual text:      فروخت اور قیمت خرید میں بالترتیب 21 پیسے اور 20
Tesseract output: فا یی نپ 2ا1
---






**Gap Analysis — Tesseract on Urdu Images (200-image dataset run)**

**Image 1: م_96.png**
Actual text: م
Tesseract output: (blank)
What went wrong: Isolated single letter — with no surrounding context, Tesseract failed to detect anything and returned a blank output.

**Image 2: 6652.png**
Actual text: والے کشمیر سیمینار کے ساتھ بھر پو تعاون کا فیصلہ
Tesseract output: الات ناک مت گر اھ
What went wrong: The entire sentence was misread — not a single word came out correctly, and word order was also lost. The newspaper scan's low contrast/noise made this especially hard for Tesseract.

**Image 3: ذ_99.png**
Actual text: ذ
Tesseract output: الا
What went wrong: Tesseract misread the isolated letter ذ as something entirely different ("الا"). Isolated characters (without any surrounding word or sentence context) keep failing — whether the output is blank or garbled.

**Image 4: Screenshot 2026-07-21 184154.png**
Actual text: ٹی ایل پی سربراہ سعد رضوی اور ان کے بھائی کی گمشدگی کا معمہ برقرار
Tesseract output: ٹی ئیل پی سریراہ سحد رضوی اور ات کے بھاکی کی گمعدگی کا مسا یرظراں... (plus extra garbage lines and stray symbols)
What went wrong: This was the closest match — several words were recognized (ٹی, پی, رضوی, بھائی کی), but nearly every word had a small spelling distortion (سربراہ → سریراہ, سعد → سحد), and extra nonsense lines/symbols were generated at the end that weren't present in the original image at all.

**Image 5: 10003.png**
Actual text: فروخت اور قیمت خرید میں بالترتیب 21 پیسے اور 20
Tesseract output: فا یی نپ 2ا1
What went wrong: Only a faint trace of the numbers survived (21 → 2ا1); the rest of the sentence came out completely garbled and unreadable.

**Summary:** This updated run on the full 200-image dataset confirms the same pattern seen in the earlier 154-image sample: (1) **isolated single letters** (م, ذ) consistently produce either a blank or garbled output, since Tesseract relies on surrounding context it doesn't have; (2) **full sentences/phrases** are either completely misread (as with 6652.png and 10003.png) or come out with mostly-recognizable words marred by small spelling errors and extra garbage lines (as with Screenshot 184154.png). This reinforces the limitations of Tesseract's pretrained Urdu model on connected/cursive script — confirming the need for a custom-trained OCR model, which will be built starting in Week 3.